misc/docs-linter/urls.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"golang.org/x/sync/errgroup"
+	"io"
+	"mvdan.cc/xurls/v2"
+	"net/http"
+	"strings"
+	"sync"
+)
+
+// extractUrls extracts urls from given file content
+func extractUrls(fileContent []byte) []string {
+	scanner := bufio.NewScanner(bytes.NewReader(fileContent))
+	urls := make([]string, 0)
+
+	// Scan file line by line
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// Extract links
+		rxStrict := xurls.Strict()
+		url := rxStrict.FindString(line)
+
+		// Check for empty links and skip them
+		if url == " " || len(url) == 0 {
+			continue
+		}
+
+		// Look for http & https only
+		if strings.HasPrefix(url, "http://") || strings.HasPrefix(url, "https://") {
+			// Ignore localhost
+			if !strings.Contains(url, "localhost") && !strings.Contains(url, "127.0.0.1") {
+				urls = append(urls, url)
+			}
+		}
+	}
+
+	return urls
+}
+
+func lintURLs(fileUrlMap map[string][]string, ctx context.Context) error {
```

Comment:
I understand this code was just moved here from somewhere else, so no need to change anything, but a few things about this function bother me.  It seems more natural to write the URLs to a channel to be processed rather than introducing a lock.  Also the context that is passed in here  is used to create the error group, but the context returned from the error group creation isn't persisted, so if the first go routine fails, it will continue to execute the rest of them. The context also isn't passed into the HTTP request so the requests will not be cancelled when the context is.  It's probably okay as is because linting is relatively short lived and performance isn't a pressing concern.  If some performance issues were to arise, it would be good to look at this more closely.